databricks/koalas/indexes.py

```diff
@@ -1962,6 +1962,56 @@ def holds_integer(self):
         """
         return isinstance(self.spark.data_type, IntegralType)
 
+    def intersection(self, other):
+        """
+        Form the intersection of two Index objects.
+
+        This returns a new Index with elements common to the index and `other`.
+
+        Parameters
+        ----------
+        other : Index or array-like
+
+        Returns
+        -------
+        intersection : Index
+
+        Examples
+        --------
+        >>> idx1 = ks.Index([1, 2, 3, 4])
+        >>> idx2 = ks.Index([3, 4, 5, 6])
+        >>> idx1.intersection(idx2).sort_values()
+        Int64Index([3, 4], dtype='int64')
+        """
+        if isinstance(other, Index):
+            if type(self) != type(other):
+                # If self Index and MultiIndex coexist on both sides,
+                # it always returns an empty MultiIndex.
+                return MultiIndex.from_tuples([(0, 0)]).take([])
+            spark_frame_other = other._internal.spark_frame.drop(NATURAL_ORDER_COLUMN_NAME)
+        elif isinstance(other, Series):
+            spark_frame_other = other.to_frame().to_spark()
+        elif isinstance(other, DataFrame):
+            return Series([])
+        elif is_list_like(other):
+            if isinstance(self, MultiIndex):
+                other = list(other)
+                spark_frame_other = MultiIndex.from_tuples(other)._internal.spark_frame.drop(
+                    NATURAL_ORDER_COLUMN_NAME
+                )
+            else:
+                spark_frame_other = Index(other)._internal.spark_frame.drop(
+                    NATURAL_ORDER_COLUMN_NAME
+                )
+        else:
+            raise TypeError("Input must be Index or array-like")
+        spark_frame_self = self._internal.spark_frame.drop(NATURAL_ORDER_COLUMN_NAME)
+        spark_frame_intersected = spark_frame_self.intersect(spark_frame_other)
```

Comment:
Also how does it work if the name of index are different?

Reply:
They just ignore the `name`.

```python
>>> pidx1 = pd.Index([1, 2, 3, 4], name="index1")
>>> pidx2 = pd.Index([3, 4, 5, 6], name="index2")

>>> pidx1.intersection(pidx2)
Int64Index([3, 4], dtype='int64')
```

Let me fix it, thanks !!